drivers/pcie/endpoint/pcie_ep_common.c

```diff
@@ -90,21 +142,149 @@ int pcie_ep_xfer_data_memcpy(const struct device *dev, uint64_t pcie_addr,
 		return mapped_size;
 	}
 
+	/*
+	 * In second attempt, we must have mapped entire size,
+	 * if not just quit here before attempting memcpy
+	 */
+	if (mapped_size != unmapped_size) {
+		pcie_ep_unmap_addr(dev, mapped_addr);
+		return -EIO;
+	}
+
 	ret = pcie_ep_mapped_copy(mapped_addr,
 				  ((uintptr_t)local_addr) + xfer_size,
 				  mapped_size, dir);
 
+	/* Flush the PCIe writes upon successful memcpy */
+	if (!ret && (dir == DEVICE_TO_HOST)) {
+		sys_read8(mapped_addr);
+	}
+
 	pcie_ep_unmap_addr(dev, mapped_addr);
 
 	/* Check if mapped_copy succeeded */
 	if (ret < 0) {
 		return ret;
 	}
 
-	/* In second attempt, we must have completed data transfer */
+	return 0;
+}
+
+/*
+ * Helper API to achieve data transfer with DMA operation through
+ * PCIe outbound memory, this API is based off pcie_ep_xfer_data_memcpy,
+ * here we use "system dma" instead of memcpy
+ */
+int pcie_ep_xfer_data_dma(const struct device *dev, uint64_t pcie_addr,
+			  uintptr_t *local_addr, uint32_t size,
+			  enum pcie_ob_mem_type ob_mem_type,
+			  enum xfer_direction dir)
+{
+	uint64_t mapped_addr;
+	int mapped_size, ret;
+	uint32_t xfer_size, unmapped_size;
+	uint32_t dummy_data;	/* For explicit dummy PCIe read */
+
+	mapped_size = pcie_ep_map_addr(dev, pcie_addr, &mapped_addr,
+				       size, ob_mem_type);
+
+	/* Check if outbound memory mapping succeeded */
```

Comment:
make the comment before line 188 (update it such as "map outbound memory"), and remove any empty lines before the if.

Reply:
ACK

Reply:
> make the comment before line 188 (update it such as "map outbound memory")

Done
> remove any empty lines before the if.

I have removed empty lines before the if, if the given if statement is checking conditions derived in the statement just above it and not for other cases in favor of readability.